### ISSUE TITLE
Machine Learning Not Possible (Token Capacity)

### ISSUE THREAD

**Issue by MohammadHomsee** [SebLague/Chess-Challenge#111]
I would like to try machine learning attempt, how can I store learning data ?

**Comment by plastic-bottle** [comment]
you aren't supposed to do full-scale machine learning - the token limit is specifically there to prevent this

**Comment by ejmejm** [comment]
It is possible to do some machine learning, as a little hint, you'll probably want to store your network in a non-numeric format. You could think about how you could represent a small NN as a string, for example. At least that is if you do it the way I am going with, but I'd imagine my way is not the only way.

**Comment by Odin7094** [comment]
There is a 1000 games to be played. I could imagine someone letting it learn on the go with some smart approach.

**Comment by Infiniti20** [comment]
> It is possible to do some machine learning, as a little hint, you'll probably want to store your network in a non-numeric format. You could think about how you could represent a small NN as a string, for example. At least that is if you do it the way I am going with, but I'd imagine my way is not the only way.

Every character in a string is counted as a token.

**Comment by ejmejm** [comment]
> > It is possible to do some machine learning, as a little hint, you'll probably want to store your network in a non-numeric format. You could think about how you could represent a small NN as a string, for example. At least that is if you do it the way I am going with, but I'd imagine my way is not the only way.
> 
> Every character in a string is counted as a token.

It is still possible with small models

**Comment by ch-iv** [comment]
I do not think that the token rule was made specifically to prevent the ML approach. ML is an allowed option (you can even see ML being mentioned as one of the possible approaches on the submission form). That being said, the token constraints make it difficult to use ML

**Comment by Firestorm-253** [comment]
> It is possible to do some machine learning, as a little hint, you'll probably want to store your network in a non-numeric format. You could think about how you could represent a small NN as a string, for example. At least that is if you do it the way I am going with, but I'd imagine my way is not the only way.

Sry, but don't even try it hahah. If u have any experience in RL u'll know that it's hard enough to make a decent chess AI without having a token limit.

**Comment by zzzzz151** [comment]
You train your NN outside, then copy paste the weights in MyBot.cs

**Comment by Firestorm-253** [comment]
> You can train your NN outside, then copy paste the weights in MyBot.cs

even then, i don't think u can do it. prove me wrong.

**Comment by MohammadHomsee** [comment]
Imagine we have neural network with 64 inputs, and 10 hidden, and one output, the weights count for this will be 640 + 10 = 650 numbers, even if I store this in one string, and even if managed to keep in token limit, The funny part that after all of this hard work, I won't be able to put my code logic in the script.

**Comment by Odin7094** [comment]
To everyone thinking about strings, if I recall correctly every character counts as one token, so just remember to check it out before implementing entire thing.

**Comment by Pds314** [comment]
You can theoretically smoosh a lot of information into 128-bit decimals which are a primitive type. How many weights or biases you can put in a decimal depends how microscopic you're willing to make your weights or biases.

**Comment by Pds314** [comment]
> You can train your NN outside, then copy paste the weights in MyBot.cs

You can theoretically do it yes. The question is can you actually come to with an efficient, powerful, and trainable architecture that can be stored and then decompressed in 1024 tokens.

**Comment by Pds314** [comment]
You could use a relatively sparse network structure and have each connection store a sort of relative address of the node to connect to and train by neuroevolution.

**Comment by NegaScout** [comment]
Neural networks are not even a good solution to a chess game (if used alone). Reinforcement learning is much better for this use. I would recommend studying the DeepBlue architecture (which internaly uses NN, but their are not the main point of it), which is based on MCTS and UCT rollouts.
